buffer/src/test/java/io/netty/buffer/api/tests/BufferAndChannelTest.java

```diff
@@ -0,0 +1,363 @@
+            assertThat(buf.writableBytes()).isEqualTo(8);
+            assertThat(channel.position()).isEqualTo(position);
+            assertThat(channel.size()).isEqualTo(size);
+        } finally {
+            channel.position(channel.size());
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("allocators")
+    public void writeMustMoveDataFromChannel(Fixture fixture) throws IOException {
+        try (BufferAllocator allocator = fixture.createAllocator();
+             Buffer buf = allocator.allocate(8)) {
+            long value = ThreadLocalRandom.current().nextLong();
+            ByteBuffer data = ByteBuffer.allocate(8).putLong(value).flip();
+            long position = channel.position();
+            assertThat(channel.write(data, position)).isEqualTo(8);
+            long size = channel.size();
+            int bytesRead = buf.writeFromChannelRead(channel, 8);
+            assertThat(bytesRead).isEqualTo(8);
+            assertThat(channel.position()).isEqualTo(8 + position);
+            assertThat(channel.size()).isEqualTo(size);
+            assertThat(buf.readableBytes()).isEqualTo(8);
+            assertEquals(value, buf.readLong());
+        } finally {
+            channel.position(channel.size());
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("allocators")
+    public void writeMustNotReadBeyondEndOfChannel(Fixture fixture) throws IOException {
+        try (BufferAllocator allocator = fixture.createAllocator();
+             Buffer buf = allocator.allocate(8)) {
+            ByteBuffer data = ByteBuffer.allocate(8).putInt(0x01020304).flip();
+            long position = channel.position();
+            assertThat(channel.write(data, position)).isEqualTo(4);
+            long size = channel.size();
+            int bytesRead = buf.writeFromChannelRead(channel, 8);
+            assertThat(bytesRead).isEqualTo(4);
+            assertThat(buf.readableBytes()).isEqualTo(4);
+            assertThat(channel.position()).isEqualTo(4 + position);
+            assertThat(channel.size()).isEqualTo(size);
+        } finally {
+            channel.position(channel.size());
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("allocators")
+    public void writeMustReturnMinusOneForEndOfStream(Fixture fixture) throws IOException {
+        try (BufferAllocator allocator = fixture.createAllocator();
+             Buffer buf = allocator.allocate(8)) {
+            long position = channel.position();
+            long size = channel.size();
+            int bytesRead = buf.writeFromChannelRead(channel, 8);
+            assertThat(bytesRead).isEqualTo(-1);
+            assertThat(buf.readableBytes()).isEqualTo(0);
+            assertThat(channel.position()).isEqualTo(position);
+            assertThat(channel.size()).isEqualTo(size);
+        } finally {
+            channel.position(channel.size());
+        }
+    }
+
+    private static FileChannel tempFileChannel() throws IOException {
```

Comment:
Ask junit for a proper tmp file location to guarantee cleanup in case test is interrupted.

Maybe worth using a mock channel here too

Reply:
Sorry for the late arrival

Reply:
For the record this is what @franz1981 is talking about: https://www.baeldung.com/junit-5-temporary-directory

Reply:
JUnit gives temp directories, not files. But I can use that to scope the files.

Mocking the channel would be putting in extra work to have less confidence that the resulting channel lives up to its interface contracts. I don't think it's worth it.